xain_fl/coordinator/coordinator.py

```diff
@@ -34,61 +34,74 @@ class Coordinator:  # pylint: disable=too-many-instance-attributes
+          :class:`~.coordinator_pb2.RendezvousRequest` and
+          :class:`~.coordinator_pb2.HeartbeatRequest`.
+
+        - ``ROUND``: A round is currently in progress. During a round
+          the important messages the coordinator can receive are
+          :class:`~.coordinator_pb2.StartTrainingRoundRequest` and
+          :class:`~.coordinator_pb2.EndTrainingRoundRequest`.  Since
+          participants are selected for rounds or not, they can be
+          advertised either ROUND or STANDBY accordingly.
+
+        - ``FINISHED``: The training session has ended and
+          participants should disconnect from the coordinator.
+
+    States are exchanged during heartbeats so that both coordinators
+    and participants can react to each others state change.
 
     The flow of the Coordinator:
+
         1. The coordinator is started and waits for enough participants to join. `STANDBY`.
         2. Once enough participants are connected the coordinator starts the rounds. `ROUND N`.
         3. Repeat step 2. for the given number of rounds
         4. The training session is over and the coordinator is ready to shutdown. `FINISHED`.
 
     Note:
+
         :class:`~.coordinator_pb2.RendezvousRequest` is always allowed
         regardless of which state the coordinator is on.
 
-    Constants:
-        DEFAULT_AGGREGATOR (:class:`~.Aggregator`): if no Aggregator instance is provided
-            during initialisation, then :class:`~.FederatedAveragingAgg` in used.
-        DEFAULT_CONTROLLER (:class:`~.Controller`): if no Controller instance is provided
-            during initialisation, then :class:`~.RandomController` in used.
-
-    Attributes:
-        num_rounds (:obj:`int`, optional): The number of rounds of the training
-            session. Defaults to 10.
-        minimum_participants_in_round (:obj:`int`, optional): The minimum number of
-            participants that participate in a round. Defaults to 1.
-        fraction_of_participants (:obj:`float`, optional): The fraction of total
-            connected participants to be selected in a single round. Defaults to 1.0,
-            meaning that all connected participants will be selected.
-            It must be in the (0.0, 1.0] interval.
-        weights (:obj:`list` of :class:`~numpy.ndarray`, optional): The weights of
-            the global model. Defaults to [].
-        epochs (:obj:`int`, optional): Number of training iterations local to
-            Participant.  Defaults to 0.
-        epochs_base (:obj:`int`, optional): Global number of epochs as of last
-            round. Defaults to 0.
-        aggregator: (:class:`~.Aggregator`, optional): The type of aggregation
-            to perform at the end of each round. Defaults to :class:`~.FederatedAveragingAgg`.
-        controller: (:class:`~.Controller`, optional): Controls how the Participants
-            are selected at the start of each round. Defaults to :class:`~.RandomController`.
-        """
+    Args:
```

Comment:
http://google.github.io/styleguide/pyguide.html#384-classes public attributes are documented as `Attributes`

Reply:
Ok, I reverted my change here.

Reply:
Actually when using `Attributes` instead of `Args`, `sphinx_autodoc_typehints` is not able to display the types of the attributes. Since args and attributes are basically the same here, I suggest we do use `Args` since it makes the docstring cleaner, so it's easier to read.

Reply:
Just reverted to using `Args` for now.

Reply:
Here is with `Args`:

![args](https://user-images.githubusercontent.com/6646324/72711583-95db3f00-3b69-11ea-9ef5-6b71fd8fc46a.png)

And with `Attributes`:

![attributes](https://user-images.githubusercontent.com/6646324/72711600-a095d400-3b69-11ea-8802-b6853bd1cd37.png)

Reply:
discussion point: should we use `Args` from now on then instead of `Attributes`?